pkg/noderesourcetopology/resourcerequests/exclusive.go

```diff
@@ -0,0 +1,82 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resourcerequests
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	v1helper "k8s.io/kubernetes/pkg/apis/core/v1/helper"
+	v1qos "k8s.io/kubernetes/pkg/apis/core/v1/helper/qos"
+)
+
+func IncludeNonNative(pod *corev1.Pod) bool {
+	for _, initContainer := range pod.Spec.InitContainers {
+		for resource := range initContainer.Resources.Requests {
+			if !v1helper.IsNativeResource(resource) {
+				return true
+			}
+		}
+	}
+	for _, container := range pod.Spec.Containers {
+		for resource := range container.Resources.Requests {
+			if !v1helper.IsNativeResource(resource) {
+				return true
+			}
+		}
+	}
+	return false
+}
+func AreExclusiveForPod(pod *corev1.Pod) bool {
+	qos := v1qos.GetPodQOS(pod)
+	return (areExclusiveForAnyContainer(qos, pod.Spec.InitContainers) ||
+		areExclusiveForAnyContainer(qos, pod.Spec.Containers))
+}
+
+func areExclusiveForAnyContainer(qos corev1.PodQOSClass, containers []corev1.Container) bool {
```

Comment:
I am wondering if something like this wouldn't be more readeable:
```suggestion
	return areExclusiveForAnyContainer(qos, append(pod.Spec.InitContainers, pod.Spec.Containers...))
}

func areExclusiveForAnyContainer(qos corev1.PodQOSClass, containers []corev1.Container) bool {
```
wdyt?

Reply:
I'm still thinking about this one (last pending comment). On one hand I don't like too much the temporary variable, on the other hand the way you suggest is pretty much the standard we use all across the codebase (including TM)

Reply:
Let's go with the flow:
```
1020 16:39:47 fromani@laptop ~/go/src/k8s.io/kubernetes $ git log --oneline -1
11067ceb328 (HEAD -> master, upstream/master, origin/master, origin/HEAD) Merge pull request #118306 from SataQiu/fix-kubeadm-20230529
1021 16:39:50 fromani@laptop ~/go/src/k8s.io/kubernetes $ git grep 'append(pod.Spec.Init'
pkg/kubelet/cm/cpumanager/cpu_manager.go:373:           for _, container := range append(pod.Spec.InitContainers, pod.Spec.Containers...) {
pkg/kubelet/cm/cpumanager/cpu_manager_test.go:183:              pod.Spec.InitContainers = append(pod.Spec.InitContainers, v1.Container{
pkg/kubelet/cm/cpumanager/policy_static.go:542: for _, container := range append(pod.Spec.InitContainers, pod.Spec.Containers...) {
pkg/kubelet/cm/cpumanager/policy_static_test.go:697:            for _, container := range append(pod.Spec.InitContainers, pod.Spec.Containers...) {
pkg/kubelet/cm/memorymanager/memory_manager.go:327:             for _, container := range append(pod.Spec.InitContainers, pod.Spec.Containers...) {
pkg/kubelet/cm/memorymanager/policy_static.go:375:      for _, ctn := range append(pod.Spec.InitContainers, pod.Spec.Containers...) {
pkg/kubelet/cm/topologymanager/scope.go:137:    for _, container := range append(pod.Spec.InitContainers, pod.Spec.Containers...) {
pkg/kubelet/cm/topologymanager/scope_container.go:53:   for _, container := range append(pod.Spec.InitContainers, pod.Spec.Containers...) {
pkg/kubelet/cm/topologymanager/scope_pod.go:60: for _, container := range append(pod.Spec.InitContainers, pod.Spec.Containers...) {
pkg/kubelet/status/status_manager_test.go:708:                  pod.Spec.InitContainers = append(pod.Spec.InitContainers, v1.Container{
plugin/pkg/admission/limitranger/admission_test.go:168:         pod.Spec.InitContainers = append(pod.Spec.InitContainers, api.Container{
plugin/pkg/admission/serviceaccount/admission.go:390:                   pod.Spec.InitContainers[i].VolumeMounts = append(pod.Spec.InitContainers[i].VolumeMounts, volumeMount)
```